### ISSUE TITLE
AutoCompleteMessages overriden to True on dotnet-isolated runtime

### ISSUE THREAD

**Issue by tony-harverson-moonpig** [Azure/azure-functions-dotnet-worker#2142]
When setting up a function using the dotnet-isolated runtime and a ServiceBusTrigger binding to a ServiceBusReceivedMessage, the servicebus "MessageAutoComplete" option is overridden to True, regardless of the host.json setting.

I've spun up a [minimal example](https://github.com/tony-harverson-moonpig/AutoCompleteTest) using the func templates.

When run locally or in Azure, the function logs contain the following:

```
 Reading host configuration file 'C:\temp\AutoCompleteTest\bin\output\host.json'
 Host configuration file read:
 {
   "version": "2.0",
   "logging": {
     "applicationInsights": {
       "samplingSettings": {
         "isEnabled": true,
         "excludedTypes": "Request"
       },
       "enableLiveMetricsFilters": true
     }
   },
   "extensions": {
     "serviceBus": {
       "autoCompleteMessages": false
     }
   }
 }
```

```
 1 functions loaded
 LoggerFilterOptions
 {
	<...Collapsed for Brevity>
 }
 LoggerFilterOptions
 {
	<...Collapsed for Brevity> 
 }
 ConcurrencyOptions
 {
	<...Collapsed for Brevity> 
 }
 FunctionResultAggregatorOptions
 {
	<...Collapsed for Brevity> 
 }
 SingletonOptions
 {
	<...Collapsed for Brevity> 
 }
 ServiceBusOptions
 {
   "ClientRetryOptions": {
     "Mode": "Exponential",
     "TryTimeout": "00:01:00",
     "Delay": "00:00:00.8000000",
     "MaxDelay": "00:01:00",
     "MaxRetries": 3
   },
   "TransportType": "AmqpTcp",
   "WebProxy": "",
   "AutoCompleteMessages": false,
   "PrefetchCount": 0,
   "MaxAutoLockRenewalDuration": "00:05:00",
   "MaxConcurrentCalls": 192,
   "MaxConcurrentSessions": 8,
   "MaxConcurrentCallsPerSession": 1,
   "MaxMessageBatchSize": 1000,
   "MinMessageBatchSize": 1,
   "MaxBatchWaitTime": "00:00:30",
   "SessionIdleTimeout": "",
   "EnableCrossEntityTransactions": false
 }
 ScaleOptions
 {
	<...Collapsed for Brevity> 
 }
 Starting JobHost
```
Then:

```
 Worker process started and initialized.
 Found the following functions:
 Host.Functions.AutoCompleteTest
 
 The 'AutoCompleteMessages' option has been overriden to 'True' value for 'Functions.AutoCompleteTest' function.
```

It's not clear to me what is overriding the AutoCompleteMessages option, or how I should stop it doing so. The In-process model features an AutoComplete trigger property, but that isn't available on the isolated model.

I'm not aware of any workaround for this issue, other that to continue using the in-process model.

**Comment by SeanFeldman** [comment]
@tony-harverson-moonpig, if you downgrade `Microsoft.Azure.Functions.Worker.Extensions.ServiceBus` a few versions back, does it work as expected or not?

**Comment by tony-harverson-moonpig** [comment]
@SeanFeldman I've tried downgrading Microsoft.Azure.Functions.Worker.Extensions.ServiceBus to:

5.12.0 (first version to support the binding to a ServiceBusReceivedMessage) - Same behaviour.
5.10.0 (editing the bind to recieve a string, rather than a ServiceBusReceivedMessage) - Same behaviour

Is there a specific version other than those it'd be helpful to try?

**Comment by SeanFeldman** [comment]
No. I was trying to eliminate an option of regression.

**Comment by asos-martinmartinov** [comment]
Do you have any update on that issue? I have the same problem. I set autoCompleteMessages as false on host.json but later function overrides it to true.

**Comment by mar2mar** [comment]
Do you have any update on that issue? I have the same problem. I set autoCompleteMessages as false on host.json but later function overrides it to true.

**Comment by SeanFeldman** [comment]
@mar2mar I'm not on the Functions team.
The issue is pending a triage.

**Comment by smackodale** [comment]
I am having the same issue, I've recently upgraded from .NET 6 (in-process) to .NET 8 Isolated and the AutoCompleteMessages field is not being respected.
The only mention of this specific log message within Azure source code is found here:
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Triggers/ServiceBusTriggerAttributeBindingProvider.cs

However this is the old webjobs library, not the new one for isolated functions.

I've tried adding the config to program.cs with no luck, it simply ignores it too:
`services.AddOptions<ServiceBusProcessorOptions>().Configure((sbOptions) => sbOptions.AutoCompleteMessages = false);`
`services.AddOptions<ServiceBusSessionProcessorOptions>().Configure((sbOptions) => sbOptions.AutoCompleteMessages = false); `

**Comment by dean-dot-git** [comment]
I've been doing some more exploring on the other options (specifically the clientRetryOptions) and it doesn't appear to be respecting any of the settings being provided.  I set the retries to something obscene... fixed, 100 max, 1 minute pauses between retries, they fire immediately.   .NET 6 In-Process respected them. 

What I'm thinking is that it isn't even reading in the host.json settings and setting it to AutoCompleteMessages to "true" as the default.

.NET Isolated SB Triggers in .NET 8 are still not there.  It's like no one tested the changes after they wrote it (or just plain forgot this critical piece).

**Comment by zmerdev** [comment]
If you throw exception at the end of the function with the service bus trigger after you've manually handled actions.CompleteMessage(msg) and DeadLetterMessage all other (non handled) messages will be sent back to the queue for reprocessing.  Bonus if you exception with a specific error code you can ignore that kind of error in your metrics until.   Pretty crap workaround but it works until this gets resolved.

**Comment by SeanFeldman** [comment]
@liliankasem, would it be possible to share any details about this issue? It's still labelled as needing triage.

**Comment by mattchenderson** [comment]
@JoshLove-msft Thoughts on this one?

**Comment by floushee** [comment]
We are in the same situation. We migrated from .NET 6 + in process model to .NET 8 and isolated. We are also relying on the _maxAutoLockRenewalDuration_ setting, which seems to be working. But no mather what we configure for the _autoCompleteMessages_, we always see the logs that AutoCompleteMessages is overwritten.

**Comment by fabiocav** [comment]
This seems to be a problem with the metadata generation causing the attribute to set this to true (which would have precedence). 

Flagging for investigation.

@floushee / @tony-harverson-moonpig we'll take a look at this in the coming sprint (starting next week) and provide an update as soon as possible.

**Comment by fabiocav** [comment]
Took an initial look at this and this might not be directly related to the metadata generation. The current worker attribute is actually missing the expected property, but that seems to be exposing another issue when that is omitted from metadata.

Exposing the property here would help (would enable the function to override) and is needed for functional parity (per function behavior configuration), but we need to review the behavior of the current WebJobs attribute.

**Comment by IuriNesterov** [comment]
Same issue here. It is so confusing, that there is all needed functionality in function app in isolated mode to handle messages manually via ServiceBusMessageActions and still messages are being completed automatically due to the fact that under the hood somewhere Azure.Functions.Worker uses Microsoft.Azure.WebJobs.ServiceBus.Triggers.ServiceBusTriggerAttributeBindingProvider which has a method that just blindly override host.json "autoCompleteMessages": false to true.

Here is the method from Microsoft.Azure.WebJobs.ServiceBus.Triggers.ServiceBusTriggerAttributeBindingProvider that does this outrageous thing:
`        private ServiceBusOptions GetServiceBusOptions(ServiceBusTriggerAttribute attribute, string functionName)
        {
            if (attribute.IsAutoCompleteOptionSet)
            {
                var options = ServiceBusOptions.DeepClone(_options);

                _logger.LogInformation($"The 'AutoComplete' option has been overrriden to '{attribute.AutoComplete}' value for '{functionName}' function.");

                options.BatchOptions.AutoComplete = options.MessageHandlerOptions.AutoComplete = options.SessionHandlerOptions.AutoComplete = attribute.AutoComplete;

                return options;
            }

            return _options;
        }`

**Comment by fabiocav** [comment]
@IuriNesterov yes, that is the logic that overrides the options value for the function. Part of the problem is how this is being handled by the attribute. 

I'm opening the PR to expose the `AutoCompleteMessages` in the worker attribute, which will give us an immediate mitigation (in addition to bringing the ability to also manage this at the function level).

**Comment by fabiocav** [comment]
The extension has been released: https://github.com/Azure/azure-functions-dotnet-worker/releases/tag/servicebus-extension-5.17.0

To mitigate the issue:

-  Upgrade to this version of the package: https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.Extensions.ServiceBus/5.17.0
- Update your code to set the `AutoCompleteMessages` property on the trigger attribute applied to your function to `false` (or what you wish to use).

Hope this helps unblock you all. Thanks for reporting and for the patience!

**Comment by SeanFeldman** [comment]
Works like a charm. Thank you, @fabiocav.

**Comment by tony-harverson-moonpig** [comment]
Thanks @fabiocav, I've tested that in the minimal example I spun up, and it works as expected, which gives us a way forward.

As far as the underlying issue goes (host.json settings being unexpectedly overridden), should we leave this issue open to track that, or should we open another issue to track that?

**Comment by fabiocav** [comment]
Thank you for the confirmation.

@tony-harverson-moonpig I left this open for now to make sure the remaining work is tracked. We might open an issue to track this in the host or SDK repo, and if that happens, I'll close this and link to that. 

Thanks again!

**Comment by mattchenderson** [comment]
I meant to provide an update here last week: the docs were updated to reflect the new attribute property and this recommendation.

**Comment by janus007** [comment]
Again.. such a basic feature that either wasn't implemented and when it finally was, it was a bug.
What is going on with this? 

I can't upgrade to 5.17 as this has another lazy bug: https://github.com/Azure/azure-functions-dotnet-worker/issues/2354 , so I downgraded, but only into this bug.

**It seems that the individuals responsible for these releases may not prioritize testing.**
I wonder if the Microsoft .NET Team is aware that people rely on coding for their livelihood.
This is not a playground; we are professionals.

**Comment by fabiocav** [comment]
Closing this as the issue tracked is now fixed.
